### ISSUE TITLE
NATS v1.4 outage due to Slow consumer issue

### ISSUE THREAD

**Issue by b1tamara** [nats-io/nats-server#1597]
Hi! 
We run cloud foundry. In the cluster we have two NATS VMs which share messages each receive so that subscribes (in our case gorouters) can connect to any instances (apps on the diego cells).
It was the issue that one VM was detected as Slow Consumer and the connection between had been closed. After that we observed that it was retried to create connection (we see a couple of times `Route Connection created` once a second).
After ~20s the NATS VM was the "Slow Consumer" again. We see again `Route Connection created`. This scenario had been repeated 6 times within 2 minutes (the NATS VM detected as Slow Consumer and the `Route Connection created`). We assume that the NATS stopped to exchange the messages (but we don't know why) as gorouters did not get the infos about routes and began to prune all of them. 
One minute after the last "Route Connection created" we observed many errors related to "TLS handshake: remote error: tls: bad certificate" logged by both of NATS VMs (related to the tls-communication with diego-cells). At the same time both of VMs CPUs increased and went up nearly 100%.

After restart of one VM, then the second one and then the first one again, NATS started to work without issues. 

We would like to understand to root cause of the issue.

**Comment by kozlovic** [comment]
On top of what @wallyqs mentioned, the fact that your TLS handshakes would start to fail (assuming TLS certificates have not changed and still valid) and the 100% CPU, it may actually be the source of the slow consumer. The VM was somehow slowed down to the point that it was causing the upstream server to close the connection because the downstream server was not dequeuing data fast enough from the socket.
You would also need to check that the 100% CPU usage was not due to a bunch of applications trying to make TLS connections to that server at that time. TLS handshake can be cpu intensive and if you had applications with wrong certs trying to (incorrectly?) connect to that server, that could explain those errors and cpu usage.

**Comment by b1tamara** [comment]
Hi! The nats.conf:
<pre><code>net: "10.0.65.3"
port: 4222
prof_port: 0
monitor_port: 0

debug: false
trace: false
logtime: true

authorization {
  user: "<value-redacted>"
  password: "<value-redacted>"
  timeout: 15
}

cluster {
  host: "10.0.65.3"
  port: 4223

  authorization {
    user: "<value-redacted>"
    password: "<value-redacted>"
    timeout: 15
  }


  tls {
    ca_file: "/var/vcap/jobs/nats/config/internal_tls/ca.pem"
    cert_file: "/var/vcap/jobs/nats/config/internal_tls/certificate.pem"
    key_file: "/var/vcap/jobs/nats/config/internal_tls/private_key.pem"
    cipher_suites: [
      "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
      "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"
    ]
    curve_preferences: [
      "CurveP384"
    ]
    timeout: 5 # seconds
    verify: true
  }


  routes = [

    nats-route://<value-redacted>:<value-redacted>@73bee12e-66ee-4d95-bc8f-e2f8bef7cbc8.nats.service.cf.internal:4223

    nats-route://<value-redacted>:<value-redacted>@8ce5aa44-6e97-48f4-a3f8-2484c7a46202.nats.service.cf.internal:4223

    nats-route://<value-redacted>:<value-redacted>@73bee12e-66ee-4d95-bc8f-e2f8bef7cbc8.nats.service.cf.internal:4225

    nats-route://<value-redacted>:<value-redacted>@8ce5aa44-6e97-48f4-a3f8-2484c7a46202.nats.service.cf.internal:4225

  ]
}</pre></code>

And the config for nats-tls
<pre><code>port: 4224
prof_port: 0
monitor_port: 0

debug: false
trace: false
logtime: true

authorization {
  user: <value-redacted>"
  password: <value-redacted>"
  timeout: 15
}

tls {
  ca_file: "/var/vcap/jobs/nats-tls/config/external_tls/ca.pem"
  cert_file: "/var/vcap/jobs/nats-tls/config/external_tls/certificate.pem"
  key_file: "/var/vcap/jobs/nats-tls/config/external_tls/private_key.pem"
  cipher_suites: [
    "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"
  ]
  curve_preferences: [
    "CurveP384"
  ]
  timeout: 5 # seconds
  verify: true
}

cluster {
  host: "10.0.65.3"
  port: 4225

  authorization {
    user: <value-redacted>"
    password: <value-redacted>"
    timeout: 15
  }


  tls {
    ca_file: "/var/vcap/jobs/nats-tls/config/internal_tls/ca.pem"
    cert_file: "/var/vcap/jobs/nats-tls/config/internal_tls/certificate.pem"
    key_file: "/var/vcap/jobs/nats-tls/config/internal_tls/private_key.pem"
    cipher_suites: [
      "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
      "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"
    ]
    curve_preferences: [
      "CurveP384"
    ]
    timeout: 5 # seconds
    verify: true
  }


  routes = [

    nats-route://<value-redacted>:<value-redacted>@73bee12e-66ee-4d95-bc8f-e2f8bef7cbc8.nats.service.cf.internal:4223

    nats-route://<value-redacted>:<value-redacted>@8ce5aa44-6e97-48f4-a3f8-2484c7a46202.nats.service.cf.internal:4223

    nats-route://<value-redacted>:<value-redacted>@73bee12e-66ee-4d95-bc8f-e2f8bef7cbc8.nats.service.cf.internal:4225

    nats-route://<value-redacted>:<value-redacted>@8ce5aa44-6e97-48f4-a3f8-2484c7a46202.nats.service.cf.internal:4225

  ]
}`</pre></code>

diego cells (450+) publish messages to the NATS. GoRouters (32) gets the messages.
 
Here you can see the monitoring. During Slow Consumer Issue the NATS Health seems to be ok. The CPU went up a little bit later 
<img width="2115" alt="92621910-b73d9680-f2c4-11ea-834f-d8841ab22b51" src="https://user-images.githubusercontent.com/34028368/93229325-c2863b80-f776-11ea-9199-ada4a4308784.png">

**Comment by b1tamara** [comment]
The slow consumer issue is related to the other NATS VM, the later "bad certificate" issue is related to the diego cells(route_emitter), which should have the valid client certificate. But we assume that the bad certificate issue is the follow-up issue. What is the other question, why route_emitters try to connect to the NATS VM again and at the same time. As I understand, the handshake between NATS VM and route_emitter runs on the "connection accept".
![Screen Shot 2020-09-15 at 11 38 54](https://user-images.githubusercontent.com/34028368/93232024-b9e33480-f779-11ea-8171-d7ce87ec5865.png)

We use cf-deployment v13.12.0, which uses nats-release v34.

**Comment by wallyqs** [comment]
Thanks for the info @b1tamara, found this issue in the CF nats-release repo that looks related: https://github.com/cloudfoundry/nats-release/issues/25

Looking at the packaging script and usage of `gnatsd` seems it is still a v1 based cluster: https://github.com/cloudfoundry/nats-release/blob/b2b5717a9cf1a9f8e3c40b4fc32c4708bab2ee9a/packages/gnatsd/packaging#L13

It would be good to start looking to upgrade to NATS Server v2 series which fix many of the conditions where Slow Consumer between routes errors appear.

**Comment by b1tamara** [comment]
Thanks for your responses. I hope that the CF community goes ahead with your comments about nats server v2

I have a further questions as we have to provide the root cause analysis for the issue. Also I investigated the nats-server coding (v1.4.1) regarding slow consumer. 

We can observe that the "Slow Consumer Detected" belongs to port 439xx as well as to default port 4223. Does it mean that the first NATS VM hosted on 10.0.65.3 tried to [flush outbound](https://github.com/nats-io/nats-server/blob/3e64f0bfd1fe4c2cf6599f064ff72fa7af439663/server/client.go#L653) to other NATS VM hosted on 10.0.129.3 listening on 43896 and detects it as Slow Consumer? But did not clear connection immediately but tried to write again and then disconnected? Does it means that the messages went lost?
The other Slow Consumer has been detected as NATS VM hosted on 10.0.65.3 was like a client and got the messages from other NATS VM hosted on 10.0.129.3 from port 4223. Does it means that the subscribes of the NATS VM were not fast enough to process the messages? Or am I wrong here and the NATS VM is slow consumer ? 

What is also not really clear why we see "Route connection created" more often as the "Slow consumer Detected" log. I understand that "Route connection created" has been log from the both sides, it means after "Slow Consumer Detected" we have to see only to logs for "Route connection created". Or had been the connections closed from both side in case of slow consumer on one side? 

Have you any idea why exact after 20 sec after route connection we see "Slow Consumer" e.g. for rid 2269. 
Another question is why two Route connection to the same ports have been created without any log messages between, e.g. at 11:43:08 and 11:43:09 connection to 10.0.129.3:4223 as well 10.0.65.3:4223...

**Comment by wallyqs** [comment]
> Does it mean that the first NATS VM hosted on 10.0.65.3 tried to flush outbound to other NATS VM hosted on 10.0.129.3 listening on 43896 and detects it as Slow Consumer? But did not clear connection immediately but tried to write again and then disconnected? Does it means that the messages went lost?

Yes that is correct, by default it allows 2s delay for consuming the data. The messages attempted to be sent on that failed write would have been lost.

>  Or am I wrong here and the NATS VM is slow consumer?

Slow Consumer error can happen on both client and route connections, but in this case it looks like the NATS VM itself was the slow consumer not the clients.

> Another question is why two Route connection to the same ports have been created without any log messages

I think this might be due to auto discovery and on reconnect another server announces same route so it attempts a connect but notices that the server id is the same so it disconnects, so there might be some for `Route connection created` entries in the logs than there are servers sometimes. Since the setup had two NATS servers then the Slow Consumer error would close the connection from both sides.

I think NATS Server v2 upgrade is a great move for CF, many performance improvements in that release and some of the logging errors are clearer.

**Comment by b1tamara** [comment]
Hi @wallyqs,

a further question. After connection was cleared / interrupted from the server side, the client did not get the messages in [readLoop](https://github.com/nats-io/nats-server/blob/3e64f0bfd1fe4c2cf6599f064ff72fa7af439663/server/client.go#L420) and the [closeConnection](https://github.com/nats-io/nats-server/blob/3e64f0bfd1fe4c2cf6599f064ff72fa7af439663/server/client.go#L1839) was executed, where all subscribers were [unsubscribed](https://github.com/nats-io/nats-server/blob/3e64f0bfd1fe4c2cf6599f064ff72fa7af439663/server/client.go#L1896). What does it mean for our subscribers? 
Did this disconnection from one NATS VM more impact as only that the NATS did not exchange the messages?

**Comment by wallyqs** [comment]
@b1tamara the impact would mostly have been that messages from node A to node B would have been lost during the reconnection, clients that were connected to the same node would have been able to receive messages.

**Comment by b1tamara** [comment]
Hello,
we try to reproduce "Slow Consumer" issue between NATS VMs on our dev system. 

In order to have a more "real-live" scenario we decided to simulate route emitters behaviour by publishing route register messages so that it would be consumed by the gorouters. We used nats-pub:

<pre><code>
git clone https://github.com/nats-io/nats.go.git  

go run nats.go/examples/nats-pub/main.go  -s nats://nats:<nats-password>@10.1.1.3:4222 router.register '{"host":"10.0.139.9","port":61063,"uris":["test-nats2-app.cf.cfi05.aws.cfi.sapcloud.io"],"app":"29449583-ba01-4982-a915-fd1597b5dd6a","private_instance_id":"3213c9e4-8cbc-435a-6f83-3faa","private_instance_index":"0","server_cert_domain_san":"3213c9e4-8cbc-435a-6f83-3faa","tags":{"app_id":"29449583-ba01-4982-a915-fd1597b5dd6a","app_name":"zfse-00052384-approuter","component":"route-emitter","instance_id":"0","organization_id":"4b705754-2fe8-4055-873c-b2915ab0124d","organization_name":"d12_daf","process_id":"29449583-ba01-4982-a915-fd1597b5dd6a","process_instance_id":"3213c9e4-8cbc-435a-6f83-3faa","process_type":"web","source_id":"29449583-ba01-4982-a915-fd1597b5dd6a","space_id":"d1d003b5-4256-4991-8e8a-15a43355941e","space_name":"SharedServices"}}'
</pre></code>

**Test scenario 1**: Try to bring one of the vms down

Stressing the CPU of one of the vms
Send a large number of route register messages using nats-pub to the same vm, sending 100000 messages over one connection
Results:
- The CPU load increased close to 100% but again kept going down and up again
- The 2 vms were up during the whole test
- No Slow Consumer between NATS VMs

**Test scenario 2**: Try to bring one of the vms down

Run nats-bench against both vms, with 100000 messages, 1000 publishers and 1000 subscribers
In parallel run nats-pub with 1000000 messages (1 connection) against one of the vms.

Results:

- The CPU usage of both instances increased ~ 100%
- First one of the instances crashed with monit summary reporting 
- NATS health of both vms is 0
- The second vm also failed 
- The VMs couldn't accept new connections
- However, the routes were still received by gorouter and **no Slow Consumer issue between NATS VMs**

**Test Scenario 3**: Stress the gorouter to reproduce the slow consumer error between gorouter and NATS VM.

Test setup:
1. Run a load test against one of the apps using vegeta:
<pre><code>echo "GET https://<url>" | vegeta -cpus=4 attack -keepalive=1 -duration=1200s -rate=1000/1s -max-body=1000 | vegeta report</pre></code>
2. Craft router.register messages (100,000) using nats-pub to be consumed by Gorouter
Results:
- Gorouter CPU load ~ 100% . However, the CPU load alone wasn't enough to hinder gorouter routes consumption rate. 
- **No Slow consumer issue between NATS VMs.**

Have you idea how we can reproduce the Slow Consumer issue? More load to NATS?

**Comment by kozlovic** [comment]
Since the "Slow Consumer" was for a route and was a write deadline (that is, TCP write blocked for more than 2 seconds), you would need to cause the network link between 2 NATS Servers to be either slow, or have the NATS server receiving data from the NATS route to be sending to consumers that are themselves slow to read from the socket, causing the server to spend a bit of time for each client connection it sends the incoming routed message to. If there lot of them, the total time spent there may be big enough that the upstream has filled the socket buffer and now blocks on a TCP write, up to the 2 seconds write deadline.

**Comment by wallyqs** [comment]
Closing this one since related issue from CF also closed, recommend to follow the recommendations from their issue: https://github.com/cloudfoundry/nats-release/issues/25
Thanks!
